cmd/deploy/wait.go

```diff
@@ -66,48 +71,56 @@ func (dw *Waiter) wait(ctx context.Context, opts *Options) error {
 	return nil
 }
 
-func (dw *Waiter) waitForResourcesToBeRunning(ctx context.Context, opts *Options) error {
-	ticker := time.NewTicker(5 * time.Second)
-	to := time.NewTicker(opts.Timeout)
+func (dw *Waiter) waitForResourcesToBeRunning(ctx context.Context, opts *Options, namespace string) error {
 	c, _, err := dw.K8sClientProvider.ProvideWithLogger(okteto.GetContext().Cfg, dw.K8sLogger)
 	if err != nil {
 		return err
 	}
 
-	for {
-		select {
-		case <-to.C:
-			return fmt.Errorf("'%s' deploy didn't finish after %s", opts.Manifest.Name, opts.Timeout.String())
-		case <-ticker.C:
-			ns := okteto.GetContext().Namespace
-			dList, err := pipeline.ListDeployments(ctx, opts.Manifest.Name, ns, c)
-			if err != nil {
-				return err
-			}
-			areAllRunning := true
-			for i := range dList {
-				d := &dList[i]
-				if !deployments.IsRunning(ctx, ns, d.Name, c) {
-					areAllRunning = false
+	// first list the pods of the pipeline and check if they are running
+	podList, err := pipeline.ListPods(ctx, namespace, opts.Manifest.Name, c)
+	if err != nil {
+		return err
+	}
+	pendingPodsMap := make(map[string]bool)
+	for i := range podList.Items {
+		p := &podList.Items[i]
+		if p.Status.Phase != v1.PodRunning {
+			pendingPodsMap[p.Name] = true
+		}
+	}
+
+	// watch if there are pending pods to be running
+	if len(pendingPodsMap) > 0 {
+		w, err := pipeline.WatchPods(ctx, namespace, opts.Manifest.Name, podList.ResourceVersion, c)
+		if err != nil {
+			return err
+		}
+		defer w.Stop()
+
+		for e := range w.ResultChan() {
+			switch e.Type {
+			case watch.Modified, watch.Added:
+				pod, ok := e.Object.(*v1.Pod)
+				if !ok {
+					continue
 				}
-			}
-			if !areAllRunning {
-				continue
-			}
-			sfsList, err := pipeline.ListStatefulsets(ctx, opts.Manifest.Name, ns, c)
-			if err != nil {
-				return err
-			}
-			for i := range sfsList {
-				ss := &sfsList[i]
-				if !statefulsets.IsRunning(ctx, ns, ss.Name, c) {
-					areAllRunning = false
+				if _, ok := pendingPodsMap[pod.Name]; ok && pod.Status.Phase == v1.PodRunning {
+					delete(pendingPodsMap, pod.Name)
 				}
-			}
-			if !areAllRunning {
+				if len(pendingPodsMap) == 0 {
+					return nil
+				}
+			// ignore this events
+			case watch.Error:
+				continue
```

Comment:
i was having doubts about this case. the idea is to keep listening until all pods are running, if this takes more than the timeout, the ctx will be canceled and it will stop listening. in that sense, for error could be interesting to log them just in case there is something wierd with the cluster, but they can also be transient?

I will think about this and upload a change

Reply:
Can we at least log in with debug in order to check what's going on if it fails?

Reply:
Moved the changes to an other PR dedicated to discuss further https://github.com/okteto/okteto/pull/4555